src/traces/pie/defaults.js

```diff
@@ -13,21 +13,52 @@ var attributes = require('./attributes');
 var handleDomainDefaults = require('../../plots/domain').defaults;
 var handleText = require('../bar/defaults').handleText;
 
-module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
+function handleLabelsAndValues(labels, values) {
+    var hasLabels = Array.isArray(labels);
+    var hasValues = Lib.isArrayOrTypedArray(values);
+    var len = Math.min(
+        hasLabels ? labels.length : Infinity,
+        hasValues ? values.length : Infinity
+    );
+
+    if(!isFinite(len)) len = 0;
+
+    if(len && hasValues) {
+        var sum = 0;
+        for(var i = 0; i < len; i++) {
+            var v = +values[i];
+            if(v < 0) {
+                sum = 0;
+                break;
+            }
+            sum += v;
+        }
+        if(!sum) len = 0;
+    }
```

Comment:
Moreover, you're potentially looping here over all the `values` items. This is something we usually defer to the calc step to keep `supplyDefaults` performance scaling as the number of attributes NOT the number of data array items. 

As `pie-like` trace usually don't have that many data array items, this is not a big deal, but still could you move some of this logic to the calc step?

Reply:
Good call. Done in https://github.com/plotly/plotly.js/pull/4416/commits/1967fd459ef07a8e09edc52a997d74539bd9914a.